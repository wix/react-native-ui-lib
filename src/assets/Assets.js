import _ from 'lodash';

function assignProperties(a, b) {
  if (a) {
    _(b).keys().forEach((key) => {
      Object.defineProperty(a, key, Object.getOwnPropertyDescriptor(b, key));
    });
  }

  return a;
}

function ensurePath(obj, path) {
  let pointer = obj;

  const pathArray = path.split('.');
  const n = pathArray.length;

  for (let i = 0; i < n; i++) {
    const segment = pathArray[i];
    
    if (pointer[segment]) {
      const descriptor = Object.getOwnPropertyDescriptor(pointer, segment);
      if (descriptor.get) {
        Object.defineProperty(pointer, segment, descriptor);
      }
    } else {
      pointer[segment] = pointer[segment] || {};
    }
    pointer = pointer[segment];
  }

  return pointer;
}

export class Assets {
  loadAssetsGroup(groupName, assets) {
    if (!_.isString(groupName)) {
      throw new Error('group name should be a string');
    }

    if (!_.isPlainObject(assets)) {
      throw new Error('assets should be a hash map or a function (for lazy access)');
    }

    if (groupName === '') {
      assignProperties(this, assets);
    } else {
      assignProperties(ensurePath(this, groupName), assets);
    }

    return this;
  }
}
