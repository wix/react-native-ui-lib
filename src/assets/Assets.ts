import _ from 'lodash';

interface CustomObject {
  [key: string]: any;
}

function assignProperties(a: CustomObject, b: {[key: string]: any}) {
  if (a) {
    Object.keys(b).forEach(key => {
      // @ts-ignore
      Object.defineProperty(a, key, Object.getOwnPropertyDescriptor(b, key));
    });
  }

  return a;
}

function ensurePath(obj: CustomObject, path: string) {
  let pointer = obj;

  const pathArray = path.split('.');
  const n = pathArray.length;

  for (let i = 0; i < n; i++) {
    const segment = pathArray[i];

    if (pointer[segment]) {
      const descriptor = Object.getOwnPropertyDescriptor(pointer, segment);
      if (descriptor?.get) {
        Object.defineProperty(pointer, segment, descriptor);
      }
    } else {
      pointer[segment] = pointer[segment] || {};
    }
    pointer = pointer[segment];
  }

  return pointer;
}

export class Assets {
  [key: string]: any;

  loadAssetsGroup(groupName: string, assets: object) {
    if (!_.isString(groupName)) {
      throw new Error('group name should be a string');
    }

    if (!_.isPlainObject(assets)) {
      throw new Error('assets should be a hash map or a function (for lazy access)');
    }

    if (groupName === '') {
      assignProperties(this, assets);
    } else {
      assignProperties(ensurePath(this, groupName), assets);
    }

    return this;
  }

  getAssetByPath(path: string) {
    return _.get(this, path);
  }
}
