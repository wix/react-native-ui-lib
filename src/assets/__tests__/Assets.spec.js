import {Assets} from '../Assets';

describe('Assets', () => {
  let assets;

  beforeEach(() => {
    assets = new Assets();
  });

  describe('.loadAssetsGroup(groupName, assets)', () => {
    it('should return the assets object itself', () => {
      expect(assets.loadAssetsGroup('emojis', {})).toBe(assets);
    });

    it('should create nested groups', () => {
      assets.loadAssetsGroup('emojis.ascii', {smile: ':)'});
      expect(assets.emojis.ascii.smile).toBe(':)');
    });

    describe('edge cases', () => {
      it('should throw if group name is not a string', () => {
        expect(() => assets.loadAssetsGroup(42, {})).toThrowErrorMatchingSnapshot();
      });

      it('should throw if assets are not a plain object', () => {
        expect(() => assets.loadAssetsGroup('assets', new class {}())).toThrowErrorMatchingSnapshot();
      });
    });

    describe('when called with empty group name', () => {
      describe('and plain object of assets', () => {
        it('should create root asset groups', () => {
          const emojis = {};
          const icons = {};
          assets.loadAssetsGroup('', {emojis, icons});

          expect(assets.emojis).toBe(emojis);
          expect(assets.icons).toBe(icons);
        });
      });

      describe('and an object with assets getters', () => {
        it('should create lazy root asset groups', () => {
          const emojis = {};
          const heavyIconsModule = {};
          const requireHeavyIcons = jest.fn().mockReturnValue(heavyIconsModule);

          assets.loadAssetsGroup('', {
            emojis,
            get icons() {
              return requireHeavyIcons();
            }
          });

          expect(assets.emojis).toBe(emojis);

          expect(requireHeavyIcons).not.toHaveBeenCalled();
          expect(assets.icons).toBe(heavyIconsModule);
          expect(requireHeavyIcons).toHaveBeenCalled();
        });
      });
    });

    describe('when has a lazy group', () => {
      let iconsModule;

      beforeEach(() => {
        iconsModule = {
          get apply() {
            return './apply.png';
          },
          get back() {
            return './back.png';
          }
        };

        assets.loadAssetsGroup('', {
          get icons() {
            return iconsModule;
          }
        });
      });

      describe('and called with the same group name', () => {
        beforeEach(() => {
          assets.loadAssetsGroup('icons', {
            get back() {
              return './back-dark.png';
            },
            get forward() {
              return './forward-dark.png';
            }
          });
        });

        it('should preserve existing assets upon load', () => {
          expect(assets.icons.apply).toBe('./apply.png');
        });

        it('should overwrite existing assets if they have the same names', () => {
          expect(assets.icons.back).toBe('./back-dark.png');
        });

        it('should append extra assets to the group', () => {
          expect(assets.icons.forward).toBe('./forward-dark.png');
        });
      });
    });
  });
});
