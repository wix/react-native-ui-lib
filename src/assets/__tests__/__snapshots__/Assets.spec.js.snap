// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assets .loadAssetsGroup(groupName, assets) edge cases should throw if assets are not a plain object 1`] = `"assets should be a hash map or a function (for lazy access)"`;

exports[`Assets .loadAssetsGroup(groupName, assets) edge cases should throw if group name is not a string 1`] = `"group name should be a string"`;
